add naru lazy brush fx

Appule opened the pull request:
# OVERVIEW
This PR adds a new effect called Naru LazyBrush.
It extracts areas enclosed by black lines and generatees a mask.
This effect is useful, for example, for cutting out characters during a line test.

It is based on the paper "LazyBrush: Flexible Painting Tool for Hand-drawn Cartoons" by Daniel Sýkora et al., EUROGRAPHICS 2009.

# How to use
Naru LazyBrush Fx generates a mask from the source image.
By default, mask markers are placed automatically, as shown in the images below:

<table align="center">
  <tr>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/8c97beda-7c87-4799-9c0a-f217446d99cd" width="300" height="600" /><br />
      <span>Source</span>
    </td>
    <td align="center">
      <img src="https://github.com/user-attachments/assets/705b5969-9a10-4c2e-8496-fb44565c956f" width="300" height="600" /><br />
      <span>Result</span>
    </td>
  </tr>
</table>



The effect also fills small gaps in the line art to some extent:

<table align="center">
  <tr>
    <td align="center">
      <img width="300" height="330" alt="rabbit" src="https://github.com/user-attachments/assets/69995d21-18d4-4e7e-b2c9-db4629301333" /><br />
      <span>Source</span>
    </td>
    <td align="center">
      <img width="300" height="330" alt="rabbit_color" src="https://github.com/user-attachments/assets/1f717924-ee53-4383-aefe-46d2a82692d1" /><br />
      <span>Result</span>
    </td>
  </tr>
</table>



You can manually place mask markers using red, and alpha markers using blue in a reference image, like this:

<table align="center">
  <tr>
    <td align="center">
      <img width="300" height="400" alt="rabbit_ref" src="https://github.com/user-attachments/assets/b8d73fc5-83ec-47f8-8dfa-f020ccea8d68" /><br />
      <span>Source + Reference</span>
    </td>
    <td align="center">
      <img width="300" height="400" alt="rabbit_ref_color" src="https://github.com/user-attachments/assets/d3f764d5-74f6-42e0-adc3-1d0c28aca3f6" /><br />
      <span>Result</span>
    </td>
  </tr>
</table>



# Parameters
- **Mode**: 
  - **Mask + Line (default)**: Generates a mask and overlays the source image.
  - **Mask**: Generates the mask only.
  - **LoG Filter**: Shows the result of the Laplacian of Gaussian filter.
  - **Capacity Map**: Displays the node capacity map used in the graph cut algorithm.
  - **Scribble Map**: Displays the map of scribbles (mask/alpha markers).

- **Mask Color**: The color used for the generated mask.
- **Fill Hole**: Controls whether fully enclosed spaces within other enclosed areas are marked as mask or alpha.
- **Auto Mask Scribble**: Automatically generates mask scribbles.
- **Line's Weights**: Enter the approximate thickness of the line art strokes in pixels.
- **Background Direction**: Decide on the general direction of the areas you want to make transparent.

**-- Advanced Parameters --**
- **Scribble Type**: Sets the type of scribble. Options: soft, hard.
- **Outline Cutoff**: Threshold for ignoring faint outlines based on brightness.
- **LoG Filter Scale**: Controls the scale of the Laplacian of Gaussian filter.
- **Filter Smoothness**: Determines how strongly ink influences the graph cut. (Lower values make it more sensitive to faint or blurry lines.)
- **Difficulty Passing Through Ink**: Adjusts how resistant ink is to being crossed by the segmentation algorithm. 
- **Scribble Strength**: Controls how strongly the scribbles influence the mask. (Lower values will correct mistakes more aggressively.)
- **Auto Scribble Margin**: Sets the margin distance from the image border where auto-scribbles will be generated.
- **Auto Scribble Outline Threshold**: Brightness threshold that triggers auto-scribbling.

# License
This Fx is based on _The Boykov-Kolmogorov Maxflow Algorithm_ from [https://github.com/Gnimuc/BKMaxflow.jl](https://github.com/Gnimuc/BKMaxflow.jl) which is published under MIT License.

RodneyBaker commented:
I'm not entirely sure how to test this as it doesn't successfully build with the automated builds.
It might be expected to build correctly via the nightlies.

konero commented:
Tested but trying to add the node causes crash.

<img width="152" height="79" alt="image" src="https://github.com/user-attachments/assets/9cdba317-807a-4d98-b8e2-a8e947009128" />

Appule commented:
Sorry, I forgot to update the CMakeLists file!

shun-iwasawa commented:
FYI:  Naru ( @Appule  ) is my colleague. He joined our studio last April. 👍

RodneyBaker commented:
Thanks for the confirmation @Appule (and welcome aboard!)
It looked like a file was missing but haven't had time to delve deeper.

<img width="988" height="773" alt="image" src="https://github.com/user-attachments/assets/eba0e8f0-4e6f-45f6-bab3-286d2048e3e8" />

Adding Shaders/FX and such has always been difficult because it is so easy to forget something.

shun-iwasawa commented:
Jenkins

Appule commented:
I’ve made it work with transparent background images. Thanks for the comment!

RodneyBaker commented:
I've been having fun with this feature.
Chasing the boundary bug a little (vector has a history of that and here it helps to have an outside boundary to prevent clipping of vector lines at the edges).  Drawing 2 strategically placed dots (or 4 dots at the corners) outside of the main drawing will do the job.

I got distracted a little by experimenting with animated scribble controls and simulating lighting by turning parts of an image mask on/off over time.

I promise I'll try to focus on the core workflow soon though!

Here's a simple example of the Camera boundary problem:

https://github.com/user-attachments/assets/7cb5de22-39f5-4b4b-993c-2b4a196ac4df

RodneyBaker commented:
@Abel032 
@shun-iwasawa has resolved the boundary problem in a few other FX so I'll defer to his work on that.  I'll see if I can find those areas of code to reference but I'm confident Shun will know.  I do agree that is likely relates to what is processed by the camera.

Appule commented:
I fixed the boundary problem. It happened because, according to the specs of the OT, I was only processing the area within the camera instead of the entire input image. With this fix, I've changed it to process the entire input image regardless of the camera's range.

By the way, I've added some new parameters: 'Line's Weights' and 'Background Direction'.

'Line's Weights' allows you to adjust where the markers are placed based on the thickness of the lines because auto-scribble didn’t work properly if the line was too thick.

'Background Direction' is a modified version of the previous 'Boundary Transparent Scribble'. It allows you to set scribbles for the background and mask areas just by specifying a rough direction (experimental).

Abel032 commented:
@RodneyBaker 
> Added: I briefly explored trying to get the code from this PR to fill all areas/shapes it encounters with different colors rather than all with the same color. I didn't get far but think it worth pursuing even if as a separate FX. Perhaps call it Naru FX Lazy Brush plus FX. ;)

I think it would be a new Check Mode called `4 Colors Check`, works together with the `Gap Check`.

Appule commented:
I fixed two miner issues since they were easy to fix.

RodneyBaker commented:
Back to having fun with lazy brush naru!

![LazyBrushNaruFX](https://github.com/user-attachments/assets/615fed4d-a303-47c6-a189-9c9c4b1e55af)

In the 'its a feature not a bug category' I am noting how rendering with a shrink value effects the results.

![LazyBrushNaruFXshrink4](https://github.com/user-attachments/assets/1cd20dd0-b574-4f8d-aca3-0db66cdfaa33)

RodneyBaker commented:
For those curious, this PR is still awaiting a review by @shun-iwasawa 
It will also need to be refreshed in light of changes to cmakelist.
The MacOS (Mojave) check can be dropped.
